openmc/mgxs_library.py

```diff
@@ -970,109 +1103,189 @@ def set_multiplicity_mgxs(self, nuscatter, scatter=None, nuclide='total',
+                                     'writing the HDF5 library')
+                xsgrp.create_dataset("nu-fission",
+                                     data=self._nu_fission[i],
+                                     compression=compression)
+
+            if self._scatter_matrix[i] is None:
+                raise ValueError('Scatter matrix must be provided when '
+                                 'writing the HDF5 library')
+            # Get the sparse scattering data to print to the library
+            G = self.energy_groups.num_groups
+            if self.representation == 'isotropic':
+                g_out_bounds = np.zeros((G, 2), dtype=np.int)
+                for g_in in range(G):
+                    nz = np.nonzero(self._scatter_matrix[i][0, g_in, :])
+                    g_out_bounds[g_in, 0] = nz[0][0]
+                    g_out_bounds[g_in, 1] = nz[0][-1]
+                # Now create the flattened scatter matrix array
+                matrix = self._scatter_matrix[i]
+                flat_scatt = []
+                for g_in in range(G):
+                    for g_out in range(g_out_bounds[g_in, 0],
+                                       g_out_bounds[g_in, 1] + 1):
+                        for l in range(len(matrix[:, g_in, g_out])):
+                            flat_scatt.append(matrix[l, g_in, g_out])
+                # And write it.
+                scatt_grp = xsgrp.create_group('scatter data')
+                scatt_grp.create_dataset("scatter matrix",
+                                         data=np.array(flat_scatt),
+                                         compression=compression)
+                # Repeat for multiplicity
+                if self._multiplicity_matrix[i] is not None:
+                    # Now create the flattened scatter matrix array
+                    matrix = self._multiplicity_matrix[i][:, :]
+                    flat_mult = []
+                    for g_in in range(G):
+                        for g_out in range(g_out_bounds[g_in, 0],
+                                           g_out_bounds[g_in, 1] + 1):
+                            flat_mult.append(matrix[g_in, g_out])
+                    scatt_grp.create_dataset("multiplicity matrix",
+                                             data=np.array(flat_mult),
+                                             compression=compression)
+                # And finally, adjust g_out_bounds for 1-based group counting
+                # and write it.
+                g_out_bounds[:, :] += 1
+                scatt_grp.create_dataset("g_min", data=g_out_bounds[:, 0],
+                                         compression=compression)
+                scatt_grp.create_dataset("g_max", data=g_out_bounds[:, 1],
+                                         compression=compression)
+
+            elif self.representation == 'angle':
+                Np = self.num_polar
+                Na = self.num_azimuthal
```

Comment:
Ahh this one I have to keep as Np (or something equally short, which I might as well just keep as is) to avoid collisions with numpy (as np).